client/consensus/grandpa/authorities.go

```diff
@@ -0,0 +1,655 @@
+			// the block that signaled the hashNumber.
+			isDesc, err := isDescendentOf(change.canonHash, bestHash)
+			if err != nil {
+				return nil, err
+			}
+			if change.canonHash == bestHash || isDesc {
+				medianLastFinalized := change.delayKind.value.(Best).medianLastFinalized
+
+				roots := authSet.pendingStandardChanges.Roots()
+				for _, standardChangeNode := range roots {
+					standardChange := standardChangeNode.change
+
+					isDescStandard, err := isDescendentOf(standardChange.canonHash, change.canonHash)
+					if err != nil {
+						return nil, err
+					}
+					if standardChange.EffectiveNumber() <= N(medianLastFinalized) && isDescStandard {
+						logger.Infof(
+							"Not applying authority set hashNumber forced at block %d, due to pending standard hashNumber at block %d",
+							change.canonHeight, standardChange.EffectiveNumber())
+						return nil, errForcedAuthoritySetChangeDependencyUnsatisfied
+					}
+				}
+
+				// apply this hashNumber: make the set canonical
+				logger.Infof("👴 Applying authority set hashNumber forced at block #%d", change.canonHeight)
+
+				// TODO telemetry
+
+				authSetChanges := authSet.authoritySetChanges
+				authSetChanges.append(authSet.setId, N(medianLastFinalized))
+				newSet = &appliedChanges[H, N]{
+					N(medianLastFinalized),
+					AuthoritySet[H, N]{
+						currentAuthorities:     change.nextAuthorities,
+						setId:                  authSet.setId + 1,
+						pendingStandardChanges: NewChangeTree[H, N](), // new set, new changes
+						pendingForcedChanges:   nil,
+						authoritySetChanges:    authSetChanges,
+					},
+				}
+				return newSet, nil
+			}
+		}
+	}
+
+	return newSet, nil
+}
+
+// applyStandardChanges Apply or prune any pending transitions based on a finality trigger. This
+// method ensures that if there are multiple changes in the same branch,
+// finalizing this block won't finalize past multiple transitions (i.e.
+// transitions must be finalized in-order). The given function
+// `is_descendent_of` should return `true` if the second hash (target) is a
+// descendent of the first hash (base).
+//
+// When the set has changed, the return value will be a Status type where newSetBlock
+// is the canonical block where the set last changed (i.e. the given
+// hash and number).
+func (authSet *AuthoritySet[H, N]) applyStandardChanges(
+	finalizedHash H,
+	finalizedNumber N,
+	isDescendentOf IsDescendentOf[H],
+	telemetry *telemetry.Client) (Status[H, N], error) {
+	// TODO telemetry here is just a place holder, replace with real
+
+	status := Status[H, N]{}
+	finalizationResult, err := authSet.pendingStandardChanges.FinalizeWithDescendentIf(&finalizedHash, finalizedNumber, isDescendentOf, func(change *PendingChange[H, N]) bool {
+		return change.EffectiveNumber() <= finalizedNumber
+	})
+	if err != nil {
+		return status, err
+	}
+
+	if finalizationResult == nil {
+		return status, nil
+	}
+
+	// Changed Case
```

Comment:
So my logic here was that I didn't think we needed the overHead of a VDT. Since all that the enum contains is changed and unchanged, where unchanged contains no data and prompts no action, I made it return a pointer where nil case represents unchanged and not nil represents change. That is why I do a check `if finalizationResult == nil {
		return status, nil
	}`
to break early in the unchanged case.

This was my reasoning, but if you think better still do just use a vdt I am more than happy to implement it

Reply:
I also came up with a comperable alternative for a VDT here https://github.com/ChainSafe/gossamer/blob/93867f96ba82b3eb71a48c0d34a86c2f06dd390d/client/consensus/grandpa/authorities.go#L574, so I would appreciate you checking it out and letting me know if I should make this a vdt as well

Reply:
I see that your version of `FinalizationResult` contains a pointer to a pending change.  And that's how you determine the `Changed` vs `Unchanged`.  If you're going to use the same name, I think it should be mimic the substrate enum. Or consider revising the forktree interfaces to return `*PendingChange` so the difference is clearer.

Reply:
Actually in this case it is the outer struct that is determining if it is `Changed` vs `Unchanged` where the inner type `*PendingChange` is used to determine if there is data associated with the change. See [here for substrate code](https://github.com/paritytech/substrate/blob/28b1f79abedbad2ec97b49277b36e1f797ca9e5d/utils/fork-tree/src/lib.rs#L708).

That being said, do you think changing name of outer struct to be more clear is sufficient, or better to create VDT to mimic the enum?

Reply:
I think if we need to hold some generic type of data for the `FinalizationResult::Changed` case, I think it makes more sense to translate the enum 1:1.